# Assignment 1, O712 Devan Assi - 400388918


# Part 1


In [ ]:
!pip install -q git+https://github.com/zhouy185/rosa-starter.git

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
#Import necessary files
import numpy as np
from starter import ZONES, TIME_BLOCKS, COSTS, PROMISE, delivery_times

In [ ]:
times = delivery_times("Far West", "Fri/Sat eve", 45)
print(len(times), "orders; average delivery time", round(times.mean(), 1),
"minutes")

210 orders; average delivery time 44.9 minutes


In [ ]:
#A)

def percentageLate(zone, time_block, promise):
  late = 0
  total = 0

#If zone is all, check all zones, otherwise only look at the one that was entered
  if zone == "all":
    zoneToCheck = ZONES
  else:
    zoneToCheck = [zone]

#If timeblock is all, check all timeblocks, otherwise only look at the one that was entered
  if time_block == "all":
    timeblockToCheck = TIME_BLOCKS
  else:
    timeblockToCheck = [time_block]

#Check the zones to check and timeblock, and call the delivery_times function to get the times
  for z in zoneToCheck:
    for t in timeblockToCheck:
      time = delivery_times(z, t, promise, 42)

#Sum the times, and the times that it is late
      total += len(time)
      late += np.sum(time > promise)

#Return late tiems as a %
  return round(late/total * 100)

#Show scenarios
print("Scenario 1, zone and time block provided: ", percentageLate("Central", "Fri/Sat eve", 45), "%")

print("Scenario 2, zone is all and time block is provided: ", percentageLate("all", "Lunch", 45), "%")

print("Scenario 3, zone is provided and time block is all: ", percentageLate("North", "all", 45), "%")

print("Scenario 4, zone and time block are all: ", percentageLate("all", "all", 45), "%")


Scenario 1, zone and time block provided:  8 %
Scenario 2, zone is all and time block is provided:  4 %
Scenario 3, zone is provided and time block is all:  8 %
Scenario 4, zone and time block are all:  9 %


In [ ]:
#B)


allZonesTimes = []

#Run a loop to get delivery times for all zone and timeblock combinations
for z in ZONES:
  for t in TIME_BLOCKS:
#Call the delivery times function and sum all the times there is a late delivery
      time = delivery_times(z, t, 45, 42)
      lateRate = np.sum(time > 45) / len(time) * 100
#Add all late results to the array
      allZonesTimes.append((z, t, lateRate))
#Sort late results by ascending order
allZonesTimes.sort(key=lambda x: x[2], reverse=True)
#iterate through array and print all results
for z, t, lateRate in allZonesTimes:
  print(f"Zone: {z}, Time Block: {t}. Late Percentage: {lateRate:.2f}%")


Zone: Far West, Time Block: Fri/Sat eve. Late Percentage: 36.67%
Zone: North, Time Block: Fri/Sat eve. Late Percentage: 18.00%
Zone: Far West, Time Block: Weekday eve. Late Percentage: 10.59%
Zone: North, Time Block: Weekday eve. Late Percentage: 9.20%
Zone: Central, Time Block: Fri/Sat eve. Late Percentage: 7.89%
Zone: Central, Time Block: Weekday eve. Late Percentage: 6.97%
Zone: Far West, Time Block: Other. Late Percentage: 6.25%
Zone: North, Time Block: Lunch. Late Percentage: 4.21%
Zone: Central, Time Block: Lunch. Late Percentage: 3.85%
Zone: North, Time Block: Other. Late Percentage: 2.50%
Zone: Far West, Time Block: Lunch. Late Percentage: 2.50%
Zone: Central, Time Block: Other. Late Percentage: 2.11%


In [ ]:
#C), was able to resuse the same function created in part A, just modify the calculation

def averageDeliveryTime(zone, time_block, promise):
  minutesTotal = 0
  total = 0

#If zone is all, check all zones, otherwise only look at the one that was entered
  if zone == "all":
    zoneToCheck = ZONES
  else:
    zoneToCheck = [zone]

#If timeblock is all, check all timeblocks, otherwise only look at the one that was entered
  if time_block == "all":
    timeblockToCheck = TIME_BLOCKS
  else:
    timeblockToCheck = [time_block]

#Check the zones to check and timeblock, and call the delivery_times function to get the times
  for z in zoneToCheck:
    for t in timeblockToCheck:
      time = delivery_times(z, t, promise, 42)

#Sum the times, and the total minutes
      total += len(time)
      minutesTotal += np.sum(time)

#Return average minutes
  return round(minutesTotal/total)

print("Scenario 1, zone and time block provided: ", averageDeliveryTime("Central", "Fri/Sat eve", 45), "min")

print("Scenario 2, zone is all and time block is provided: ", averageDeliveryTime("all", "Lunch", 45), "min")

print("Scenario 3, zone is provided and time block is all: ", averageDeliveryTime("North", "all", 45), "min")

print("Scenario 4, zone and time block are all: ", averageDeliveryTime("all", "all", 45), "min")


Scenario 1, zone and time block provided:  32 min
Scenario 2, zone is all and time block is provided:  28 min
Scenario 3, zone is provided and time block is all:  31 min
Scenario 4, zone and time block are all:  32 min


In [ ]:
#D)

allZonesAverage = []

#Run a loop to get delivery times for all zone and timeblock combinations
for z in ZONES:
  for t in TIME_BLOCKS:
#Call the delivery times function and calculate the averages
      time = delivery_times(z, t, 45, 42)
      totalDeliveries = len(time)
      averageRate = np.sum(time)/totalDeliveries

#Add all average results to the array
      allZonesAverage.append((z, t, averageRate))
#Sort average results by ascending order
allZonesAverage.sort(key=lambda x: x[2], reverse=True)
#iterate through array and print all results
for z, t, averageRate in allZonesAverage:
  print(f"Zone: {z}, Time Block: {t}. Average Delivery Time: {averageRate:.2f} min")

Zone: Far West, Time Block: Fri/Sat eve. Average Delivery Time: 42.98 min
Zone: North, Time Block: Fri/Sat eve. Average Delivery Time: 37.07 min
Zone: Far West, Time Block: Weekday eve. Average Delivery Time: 35.25 min
Zone: Central, Time Block: Fri/Sat eve. Average Delivery Time: 32.39 min
Zone: North, Time Block: Weekday eve. Average Delivery Time: 31.94 min
Zone: Far West, Time Block: Lunch. Average Delivery Time: 30.37 min
Zone: Far West, Time Block: Other. Average Delivery Time: 30.33 min
Zone: Central, Time Block: Weekday eve. Average Delivery Time: 29.98 min
Zone: North, Time Block: Lunch. Average Delivery Time: 28.91 min
Zone: North, Time Block: Other. Average Delivery Time: 27.94 min
Zone: Central, Time Block: Lunch. Average Delivery Time: 27.32 min
Zone: Central, Time Block: Other. Average Delivery Time: 26.24 min


E)

In Rosas situation, it is most likely that B) the average amount of late orders is more relevant to her business.

From the results in D) we can see that every single zone and average combo was below the 45 minute promise window, with the highest being 42.98 minutes. From this information, we know that there will be some orders that are over the promise time, but we don't know by how much in orer to address it.

When we look at the data calculated in B), it says that the worst zone has 36.67% late orders, and the next closest is 18%. This gives hard data that the worst zone is almost 18% worse than any other, allowing Rosa to take action accordingly.

If she relied on the data from D), where the highest average was 42.98 minutes and the closest to that is 37.07 minutes, there is no way to truly tell how many late orders are produced by the approx 5 minute difference.

# Part 2

In [ ]:
#A

#Define new function to determine late cost per order
def costPerLateOrder(costs):
  #Calculate late cost by pulling data from dictionary
  lateCost = costs['refund'] + costs['margin']*costs['churn_orders']
  return lateCost

print(costPerLateOrder(COSTS))

26.2


Part B

I believe that the best way to choose the range would be to check in small intervals. This way we can see at which point do we reach the maximum, and can have a good degree of granularity within the readings. The range I will go with will be by checking in 3 minute intervals, and going to 60 minutes.

This range can effectively capture the low end and higher end while being able to locate the optimal point.

In [ ]:
#B

promiseTimes = [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60]

def costPerPromiseTime(zone, time_block, promiseTimes, costs):

  bestProfit = 0
  bestPromiseTime = 0
  bestZone = 0
  bestTimeBlock = 0
    #If zone is all, check all zones, otherwise only look at the one that was entered
  if zone == "all":
    zoneToCheck = ZONES
  else:
    zoneToCheck = [zone]

#If timeblock is all, check all timeblocks, otherwise only look at the one that was entered
  if time_block == "all":
    timeblockToCheck = TIME_BLOCKS
  else:
    timeblockToCheck = [time_block]

#Check the zones to check and timeblock, and call the delivery_times function to get the times
  for z in zoneToCheck:
    for t in timeblockToCheck:
      for p in promiseTimes:
        late = 0
        total = 0
        time = delivery_times(z, t, p, 42)
        total += len(time)
        late += np.sum(time > p)
        totalProfit = total*costs['margin'] - late*costPerLateOrder(costs)
        if totalProfit > bestProfit:
          bestProfit = totalProfit
          bestPromiseTime = p
          bestZone = z
          bestTimeBlock = t

  return print("For the", bestZone, "zone, and the Time Block", bestTimeBlock, ", the Best profit is: $", round(bestProfit, 2), ", and the optimal promise time is:", bestPromiseTime, "min.")

costPerPromiseTime("Central", "Fri/Sat eve", promiseTimes, COSTS)




For the Central zone, and the Time Block Fri/Sat eve , the Best profit is: $ 1401.4 , and the optimal promise time is: 48 min.
